src/arbitrary/vanchor_data/mod.rs

```diff
@@ -0,0 +1,192 @@
+use ark_ff::vec::Vec;
+use ethabi::{encode, Token};
+use tiny_keccak::{Hasher, Keccak};
+#[derive(Clone)]
+pub struct VAnchorArbitraryData {
+	pub recipient: Token,
+	pub ext_amount: Token,
+	pub relayer: Token,
+	pub fee: Token,
+	pub encrypted_output1: Token,
+	pub encrypted_output2: Token,
+}
+
+impl VAnchorArbitraryData {
+	pub fn new(
+		recipient: Token,
+		ext_amount: Token,
+		relayer: Token,
+		fee: Token,
+		encrypted_output1: Token,
+		encrypted_output2: Token,
+	) -> Self {
+		VAnchorArbitraryData::check_inputs(
+			&recipient,
+			&ext_amount,
+			&relayer,
+			&fee,
+			&encrypted_output1,
+			&encrypted_output2,
+		);
+		Self {
+			recipient,
+			ext_amount,
+			relayer,
+			fee,
+			encrypted_output1,
+			encrypted_output2,
+		}
+	}
+
+	pub fn hash_data(
+		recipient: Token,
+		ext_amount: Token,
+		relayer: Token,
+		fee: Token,
+		encrypted_output1: Token,
+		encrypted_output2: Token,
+	) -> Vec<u8> {
+		VAnchorArbitraryData::check_inputs(
+			&recipient,
+			&ext_amount,
+			&relayer,
+			&fee,
+			&encrypted_output1,
+			&encrypted_output2,
+		);
+		let tuple = [Token::Tuple(vec![
+			recipient,
+			ext_amount,
+			relayer,
+			fee,
+			encrypted_output1,
+			encrypted_output2,
+		])];
+		let encoded_input = encode(&tuple);
+		let bytes: &[u8] = &encoded_input;
+		let mut hasher = Keccak::v256();
+		hasher.update(bytes);
+		let mut res: [u8; 32] = [0; 32];
+		hasher.finalize(&mut res);
+		res.to_vec()
+	}
+
+	pub fn hash_data_self(&self) -> Vec<u8> {
```

Comment:
```suggestion
	pub fn hash(&self) -> Vec<u8> {
```

A better name for the self hash is just to `hash` itself.